File: client/src/types/order.ts
import type { Product, UnavailableProduct } from './product'

interface OrderItem {
  id: string
  product: Product
  amount: number
  price: number
}

export interface Order {
  id: string
  price: number
  received: string
  status: string
  expectedDelivery: string | null
  items: OrderItem[]
}

export interface OngoingOrder {
  totalPrice: number
  items: OrderItem[]
  unavailableProducts?: UnavailableProduct[]
}

export interface PlacedOrder {
  id: string
  price: number
  status: string
  received: string
  expectedDelivery: string | null
  items: OrderItem[]
}

export interface UserOrder {
  id: string
  userEmail: string
  price: number
  status: string
  received: string
  expectedDelivery: string | null
  items: OrderItem[]
}

export enum OrderStatus {
  PENDING,
  SENT
}

export function orderStatusToString(status: OrderStatus): string {
  switch (status) {
    case OrderStatus.PENDING:
      return 'PENDING'
    case OrderStatus.SENT:
      return 'SENT'
    default:
      return 'INVALID_STATUS'
  }
}

export enum DeliveryMethod {
  HOME_DELIVERY
}

export function deliveryMethodToString(method: DeliveryMethod): string {
  switch (method) {
    case DeliveryMethod.HOME_DELIVERY:
      return 'HOME_DELIVERY'
    default:
      return 'INVALID_DELIVERY_METHOD'
  }
}

export enum PaymentMethod {
  PAY_ON_DELIVERY
}

export function paymentMethodToString(method: PaymentMethod): string {
  switch (method) {
    case PaymentMethod.PAY_ON_DELIVERY:
      return 'PAY_ON_DELIVERY'
    default:
      return 'INVALID_PAYMENT_METHOD'
  }
}

File: server/src/main/java/me/code/springboot_neo4j/models/nodes/Order.java
package me.code.springboot_neo4j.models.nodes;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.neo4j.core.schema.GeneratedValue;
import org.springframework.data.neo4j.core.schema.Id;
import org.springframework.data.neo4j.core.schema.Node;
import org.springframework.data.neo4j.core.schema.Relationship;
import org.springframework.data.neo4j.core.support.UUIDStringGenerator;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@Node("Order")
public class Order {

    @Id
    @GeneratedValue(UUIDStringGenerator.class)
    private String id;
    private Status status;
    private double price;
    private PaymentMethod paymentMethod;
    private float latitude;
    private float longitude;
    private DeliveryMethod deliveryMethod;
    private LocalDateTime received;
    private LocalDateTime expectedDelivery;

    @Relationship(type = "PLACED_BY")
    User user;

    @Relationship(type = "INCLUDES")
    private List<OrderItem> items;

    public Order(
            User user,
            List<OrderItem> items,
            float latitude,
            float longitude,
            DeliveryMethod deliveryMethod,
            PaymentMethod paymentMethod) {
        this.status = Status.PENDING;
        this.user = user;
        this.items = items;
        this.price = getTotalPrice();
        this.latitude = latitude;
        this.longitude = longitude;
        this.deliveryMethod = deliveryMethod;
        this.paymentMethod = paymentMethod;
        this.received = LocalDateTime.now();
        this.expectedDelivery = null;
    }

    public double getTotalPrice() {
        return formatPrice(items.stream()
                .mapToDouble(OrderItem::getPrice)
                .sum());
    }

    private double formatPrice(double price) {
        return Math.round(price * 100.0) / 100.0;
    }

    public enum Status {
        PENDING, SENT
    }

    public enum DeliveryMethod {
        DRONE_DELIVERY
    }

    public enum PaymentMethod {
        PAY_ON_DELIVERY
    }
}
